org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/LinkResourceUtil.java

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat, Inc. and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.managers;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.BasicFileAttributeView;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.stream.Stream;
+
+import org.eclipse.core.internal.preferences.EclipsePreferences;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IProjectDescription;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.core.IJavaProject;
+
+class LinkResourceUtil {
+
+	private static final IPath METADATA_FOLDER_PATH = ResourcesPlugin.getWorkspace().getRoot().getLocation().append(".projects");
+
+	private static boolean isNewer(Path file, Instant instant) throws CoreException {
+		try {
+			Instant creationInstant = Files.getFileAttributeView(file, BasicFileAttributeView.class).readAttributes().creationTime().toInstant();
+			// match accuracy
+			ChronoUnit smallestSupportedUnit = Stream.of(ChronoUnit.NANOS, ChronoUnit.MILLIS, ChronoUnit.SECONDS) // IMPORTANT: keeps units in Stream from finer to coarser
+				.filter(creationInstant::isSupported).filter(instant::isSupported) //
+				.findFirst().orElse(null);
+			if (smallestSupportedUnit != null) {
```

Comment:
Check on MacOS today, looks like we still have the problem here.

`smallestSupportedUnit` is assigned to `ChronoUnit.NANOS`, though the accuracy of `creationInstant` is second, makes this method returns `false` finally.

Reply:
OK thanks, so should we just hardcode the rule for macOS to trim to the second?

Reply:
I guess so. I couldn't think of any better solution for this case.

(I was thinking about find out all the already existing metadata files before import started, but that may have too much perf penalty and may be not worth it.)

Reply:
> I was thinking about find out all the already existing metadata files before import started, but that may have too much perf penalty and may be not worth it.

The difficulty here is that we don't know at that stage what folders the importers are going to consider as projects, so we're not sure which files matter or not.

Reply:
I guess you can just those files start from `rootPaths`, because all the metadata files that exist before import happens will not be linked, so we don't need to care about what folders will be imported by which importers.

I'm not optimistic about the performance though...

Reply:
You mean recursively crawling the whole subtrees under rootPAths for pre-existing `.classpath`, `.project` and so on?
Indeed, it could impact heavily performance for users using the wrong root (eg their home or filesystem root), but for a regular development project, it should be fast enough (I don't expect it to take more than 1 seconds for the biggest projects). But if we're ready to loose 1 second, we can just put a 1 second wait directly after `Instant.now` , and the check will then be reliable on all OSs ;)

Reply:
In the new patch, I've hardcoded that Instant should be truncated and compared by the second on macOS.

Reply:
> But if we're ready to loose 1 second, we can just put a 1 second wait directly after Instant.now , and the check will then be reliable on all OSs

That's a good insight! I think there are some key metrics we need to track when roll out this change. Then we can make the decision based on the data, like:
- how many users happen to meet the two equally instants on second during import.
- how many users fail to link the files.

For example, if the data shows that very few users have two equally instant on second, then we can just simply trim to the second.